src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/AvoidOutParameters.cs

```diff
@@ -0,0 +1,109 @@
+                                                                             DiagnosticCategory.Design,
+                                                                             DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                             isEnabledByDefault: false,
+                                                                             description: s_localizableDescription,
+                                                                             helpLinkUri: "https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1021",
+                                                                             customTags: FxCopWellKnownDiagnosticTags.PortedFxCopRule);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(csac =>
+            {
+                var outAttributeType = csac.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesOutAttribute);
+                var analyzer = new MethodAnalyzer(outAttributeType);
+
+                csac.RegisterSymbolAction(analyzer.Analyze, SymbolKind.Method);
+            });
+        }
+
+        private class MethodAnalyzer
+        {
+            private readonly INamedTypeSymbol? outAttributeSymbol;
+
+            public MethodAnalyzer(INamedTypeSymbol? outAttributeSymbol)
+            {
+                this.outAttributeSymbol = outAttributeSymbol;
+            }
+
+            public void Analyze(SymbolAnalysisContext analysisContext)
+            {
+                var methodSymbol = (IMethodSymbol)analysisContext.Symbol;
+
+                if (!methodSymbol.MatchesConfiguredVisibility(analysisContext.Options, Rule, analysisContext.CancellationToken))
+                {
+                    return;
+                }
+
+                var numberOfOutParams = methodSymbol.Parameters.Count(IsOutParameter);
+
+                if (numberOfOutParams >= 1 &&
+                    !IsTryPatternMethod(methodSymbol, numberOfOutParams) &&
+                    !IsDeconstructPattern(methodSymbol))
+                {
+                    analysisContext.ReportDiagnostic(methodSymbol.CreateDiagnostic(Rule));
+                }
+            }
+
+            private bool IsOutParameter(IParameterSymbol parameterSymbol) =>
+                parameterSymbol.RefKind == RefKind.Out ||
+                // Handle VB.NET special case for out parameters
+                (parameterSymbol.RefKind == RefKind.Ref && parameterSymbol.HasAttribute(outAttributeSymbol));
```

Comment:
If there isn't a better way to handle this case, I think that is might be worth having this extension method directly available in Roslyn, WDYT?

Reply:
Yes, might be worth filing an API request in Roslyn.